crs_perception/src/model_to_point_cloud.cpp

```diff
@@ -0,0 +1,125 @@
+#include <crs_perception/model_to_point_cloud.hpp>
+
+#include <pcl/filters/voxel_grid.h>
+#include <vtkCellArray.h>
+#include <vtkPolyDataMapper.h>
+#include <vtkTriangle.h>
+#include <vtkTriangleFilter.h>
+
+namespace crs_perception
+{
+  bool ModelToPointCloud::convertToPCL(pcl::PointCloud<pcl::PointXYZ>::Ptr &point_cloud)
+  {
+    // todo(ayoungs): handle multiple file formats?
+    // todo(ayoungs): handle error for bad file
+    vtkSmartPointer<vtkPolyData> polydata1 = vtkSmartPointer<vtkPolyData>::New();
+    vtkSmartPointer<vtkOBJReader> readerQuery = vtkSmartPointer<vtkOBJReader>::New();
+    readerQuery->SetFileName(file_path_.c_str());
+    readerQuery->Update();
+    polydata1 = readerQuery->GetOutput();
+
+    //make sure that the polygons are triangles!
+    vtkSmartPointer<vtkTriangleFilter> triangleFilter = vtkSmartPointer<vtkTriangleFilter>::New();
+    triangleFilter->SetInputData(polydata1);
+    triangleFilter->Update();
+
+    vtkSmartPointer<vtkPolyDataMapper> triangleMapper = vtkSmartPointer<vtkPolyDataMapper>::New();
+    triangleMapper->SetInputConnection(triangleFilter->GetOutputPort());
+    triangleMapper->Update();
+    polydata1 = triangleMapper->GetInput();
+
+    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud = boost::make_shared<pcl::PointCloud<pcl::PointXYZ> >();
+    uniform_sampling(polydata1, num_samples_, *cloud);
+
+    // Voxelgrid
+    pcl::VoxelGrid<pcl::PointXYZ> grid_;
+    grid_.setInputCloud(cloud);
+    grid_.setLeafSize(leaf_size_, leaf_size_, leaf_size_);
+
+    grid_.filter(*point_cloud);
```

Comment:
You can directly load a point cloud using pcl::io, why go through VTK or use a voxel grid?

Reply:
As far as I'm aware, PCL can load an STL/PLY/OBJ file into a PolygonMesh object and can convert that to a pointcloud object, but it simply just uses the vertices of the polygon. I did not think this would be sufficient to convert the model into a pointcloud this way. The method I found with VTK actually samples points on the surface to create a point cloud. I thought this would be a better point cloud to perform ICP against.